types/mongoose/index.d.ts

```diff
@@ -983,13 +899,26 @@ declare module "mongoose" {
     }
 
     interface ValidateFn<T> {
-      (obj: RegExp | Function, message?: string, type?: string): T;
+      (value: T): boolean;
+    }
+
+    interface AsyncValidateFn<T> {
+      (value: T, done: (result: boolean) => void): void;
+    }
+
+    interface ValidateOptsBase {
+      msg?: string;
+      type?: string;
     }
 
-    interface ValidateOpts {
-      validator?: RegExp | Function,
-      msg?: string,
-      type?: string
+    interface ValidateOpts extends ValidateOptsBase {
+      isAsync?: false;
+      validator?: RegExp | ValidateFn<any>;
+    }
+
+    interface AsyncValidateOpts extends ValidateOptsBase {
+      isAsync: true;
```

Comment:
How come these are required when the fields in `ValidateOpts` are optional?

Reply:
See `ValidateOpts` discuttion above